capstone_project/src/main/java/com/google/sps/data/PostService.java

```diff
@@ -103,16 +119,78 @@ public void storePost(HttpServletRequest request) {
    * the new upvote count after increase.
    */
   public long upvotePost(HttpServletRequest request) {
-    Query query = new Query("Post");
+    long postID = Long.parseLong(request.getParameter("id"));
+    Key key = KeyFactory.createKey("Post", postID);
+    Entity postToUpvote = null;
 
-    List<Entity> posts = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+    try {
+      postToUpvote = datastore.get(key);
+    } catch (EntityNotFoundException ok) {
+      // this is okay because by default postToUpvote is null and there is a null check
+    }
 
-    int postID = Integer.parseInt(request.getParameter("id"));
-    Entity postToUpvote = posts.get(postID);
+    if (postToUpvote == null) {
+      return -1;
+    }
 
     long currentUpvotes = (long) postToUpvote.getProperty("upvotes");
     postToUpvote.setProperty("upvotes", ++currentUpvotes);
     datastore.put(postToUpvote);
     return currentUpvotes;
   }
+
+  /**
+   * Sorts a list of entities based on the sort type passed in and returns the sorted list. If the
+   * sort type does not exist, an empty list is returned.
+   */
+  public List<Entity> sortEntities(String sortType, List<Entity> entities) {
+    initializeSorters();
```

Comment:
shouldn't initializeSorters() be in the constructor?

Reply:
Yea. Yesterday I had it that way but it wasn't possible to mock the clock due to method references. The method references still referenced the original clock rather than the mocked clock, even though the class was referencing the mocked clock everywhere else (i think because the sorters were initialized before the mocked clock was set). I chose to go with a builder since it allows to set properties without requiring them in a constructor and method references can point to the mocked clock. This also let me move the initializeSorters method into the constructor